lib/diskoverflow/slice.go

```diff
@@ -0,0 +1,169 @@
+// Copyright (C) 2018 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package diskoverflow
+
+import (
+	"encoding/binary"
+)
+
+type Slice interface {
+	Common
+	Append(v Value)
+	Bytes() int64 // Total estimated size of contents
+	Iter(fn func(v Value) bool, rev bool)
+	IterAndClose(fn func(v Value) bool, rev bool)
+}
+
+func NewSlice(location string) Slice {
+	s := &slice{
+		key:      lim.register(),
+		location: location,
+	}
+	s.commonSlice = &memorySlice{key: s.key}
+	return s
+}
+
+type commonSlice interface {
+	common
+	append(v Value)
+	bytes() int64
+	iter(fn func(v Value) bool, rev bool, closing bool)
+}
+
+type slice struct {
+	commonSlice
+	inactive commonSlice
+	location string
+	key      int
+	spilling bool
+}
+
+func (s *slice) Append(v Value) {
+	if !s.spilling && lim.add(s.key, v.Bytes()) {
+		s.inactive = s.commonSlice
+		s.commonSlice = &diskSlice{&diskSorted{diskMap: newDiskMap(s.location)}}
+	}
+	s.append(v)
+}
+
+func (s *slice) Bytes() int64 {
+	if s.spilling {
+		return s.bytes() + lim.bytes(s.key)
+	}
+	return lim.bytes(s.key)
+}
+
+func (s *slice) Close() {
+	s.close()
+	if s.spilling {
+		s.inactive.close()
+	}
+	lim.deregister(s.key)
+}
+
+func (s *slice) Iter(fn func(v Value) bool, rev bool) {
+	s.iterImpl(fn, rev, false)
+}
+
+func (s *slice) IterAndClose(fn func(Value) bool, rev bool) {
+	s.iterImpl(fn, rev, true)
+	s.Close()
+}
+
+func (s *slice) iterImpl(fn func(Value) bool, rev, closing bool) {
+	if !s.spilling {
+		s.iter(fn, rev, closing)
+		return
+	}
+	if rev {
+		s.iter(fn, true, closing)
+		s.inactive.iter(fn, true, closing)
+		return
+	}
+	s.inactive.iter(fn, false, closing)
+	s.iter(fn, false, closing)
+}
+
+func (s *slice) Length() int {
+	if !s.spilling {
+		return s.length()
+	}
+	return s.length() + s.inactive.length()
+}
+
+type memorySlice struct {
+	key    int
+	values []Value
+}
+
+func (s *memorySlice) append(v Value) {
+	s.values = append(s.values, v)
+}
+
+func (s *memorySlice) bytes() int64 {
+	return lim.bytes(s.key)
+}
+
+func (s *memorySlice) close() {
+	s.values = nil
+}
+
+func (s *memorySlice) iter(fn func(Value) bool, rev, closing bool) {
+	if closing {
+		defer s.close()
+	}
+	orig := s.bytes()
+	removed := int64(0)
+	for i := 0; i < len(s.values); i++ {
+		if rev {
+			i = len(s.values) - 1 - i
+		}
+		if !fn(s.values[i]) {
+			return
+		}
+		if closing {
```

Comment:
I guess it would make sense if we were somehow appending as we iterate, but we never do that, so doing all of this stuff potentially slows things down, and doesn't alleviate memory pressure as I suspect it's unlikely for GC to be invoked while while we are dereferencing things.

Reply:
> [...] as I suspect it's unlikely for GC to be invoked while while we are dereferencing things.

E.g. scanning can take a long time, where I'd expect GC to kick in. Also huge sync operations (e.g. resolving conflicts on initial sync) do take a lot of time on a "normal" laptop. And all of this only happens when the container is bigger than 10MB.